### ISSUE TITLE
Trained Agent not performing as wel as TensorBoard claims...

### ISSUE THREAD

**Issue by tr1pzz** [openai/universe-starter-agent#56]
I've been playing around with the example code for a few days now but I keep getting the same issue:
eg for Pong, after about 3 hours of training, the TensorBoard global/episode_reward goes to about +20, so I'm assuming that means the agent wins with on average 20 points.

Now, I have changed the saver in worker.py to also dump the meta_graph and added some ops from model.py to a tf_collection. Next, I've created the following IPython notebook to visualise the trained Agent: 

```
import numpy as np
import gym
import cv2
import tensorflow as tf
import time
import os
import scipy.misc

def _process_frame42(frame):
    frame = frame[34:34+160, :160]
    frame = cv2.resize(frame, (80, 80))
    frame = cv2.resize(frame, (42, 42))
    frame = frame.mean(2)
    frame = frame.astype(np.float32)
    frame *= (1.0 / 255.0)
    frame = np.reshape(frame, [42, 42, 1])
    return frame

def softmax(x):
    e_x = np.exp(x - np.max(x))
    return e_x / np.sum(e_x)

def categorical_sample(x,logistic = True):
    softmax_logits = softmax(x)
    if logistic:
        return np.argmax(np.random.multinomial(1, softmax_logits))
    else:
        return np.argmax(x)

folder = "/home/tr1pzz/Desktop/Universe/universe_dumps/logs2/visualise"
games_to_play = 200
games_to_play_before_rendering = 5

games_played = 0
total_episode_reward = 0

all_episode_rewards = []
env = gym.make("Pong-v0")
#env = gym.make("Breakout-v0")
input_frame = [_process_frame42(env.reset())]

for filename in os.listdir(folder):
    if "ckpt" in filename and "meta" not in filename:
        save_path = folder+"/"+filename
print("Model will be loaded from: %s" %save_path)

initial_state = np.zeros((1,256)).astype(float)
prev_state = [initial_state, initial_state]

tf.reset_default_graph()

with tf.Session(config=tf.ConfigProto(allow_soft_placement=True, log_device_placement=True)) as sess:
    new_saver = tf.train.import_meta_graph(save_path+".meta",clear_devices=True)
    new_saver.restore(sess, save_path)
    
    inference_op = tf.get_collection("inference_logits")[0]
    state_out_0 = tf.get_collection("state_out_0")[0]
    state_out_1 = tf.get_collection("state_out_1")[0]
    get_sample_op = tf.get_collection("action_sample")[0]
    
    get_output_state = [state_out_0, state_out_1]
    print("Model loaded, ready for simulation!")
    
    while games_played<games_to_play:
        if(games_played>games_to_play_before_rendering):
            env.render()
        #time.sleep(0.003) 
        feed_dict = {"global/input_pixels:0": input_frame, 
                     "global/LSTM_cin:0": prev_state[0], 
                     "global/LSTM_hin:0": prev_state[1]}
        
        logits, state, sampled_action = sess.run([inference_op, get_output_state, get_sample_op], feed_dict = feed_dict)
        #action = categorical_sample(logits[0])
        action_tf = sampled_action.argmax()
        #print(action,action_tf)
        
        observation, reward, done, info = env.step(action_tf)
        total_episode_reward += reward
        input_frame = [_process_frame42(observation)]
        prev_state = state
        
        if done:
            print("Game done. Score was %d" %total_episode_reward)
            all_episode_rewards.append(total_episode_reward)
            observation = env.reset()
            input_frame = [_process_frame42(observation)]
            total_episode_reward = 0
            games_played += 1
```

The problem is that when I run the trained agent it loses most of it's games, averaging a score of something like -14... Any idea what I am doing wrong? Initially I thought it might have something to do with the LSTM state needing to 'warm up' but that doesn't seem to be the case..

Additionally, the first time I ran the code with 7 workers it was running at about 45fps per thread, now I only get half that speed running on the same (local) laptop.. Any ideas as to what might cause this to happen? Finally, is there an easier way to run a trained agent, or is what I'm doing the general way to approach this?

Thanks in advance!

**Comment by tlbtlbtlb** [comment]
Cool idea, trying to use just the meta graph in a standalone program! 

I think the general idea should work. Some detail must be wrong. The action sampling looks suspect:
```
        action_tf = sampled_action.argmax()
```
It's not clear where the nondeterminism comes from. If you change the policy from stochastic to deterministic, it may or may not work. For example, the agent may have learned to keep the paddle in one place by setting (center=0%, up=50%, down=50%) rather than (center=100%, up=0%, down=0%).

About speed: calling `env.render()` syncs to the monitor's frame rate. Usually that's 60 Hz, but some combination of high-res monitors and MacBooks change it to 30 Hz.

**Comment by tr1pzz** [comment]
action_tf is already nondeterministic since it comes from get_sample_op, which refers to self.sample from the LSTMPolicy in model.py.. (I had also thought about that, thats why you can see an outcommented line #action = ... where I implemented the nondeterminism myself using numpy (see functions at top of script)).

So basically I have no idea why the trained Agent could be performing differently than what TensorBoard is claiming...

And for speed I was actually referring to training (where I'm obviously not rendering the environment) and I have never changed devices..

**Comment by tlbtlbtlb** [comment]
## Determinism

I see, that makes sense.

I don't think your code resets the LSTM state at the end of episodes. The trainer does, here: https://github.com/openai/universe-starter-agent/blob/master/a3c.py#L147. That might make a big difference.

If that doesn't fix it, maybe print out the values of observation, lstm_c, lstm_h, action_logits, and action both in the trainer and your code, and see if you see differences. The environment is only slightly randomized, so you could hope to see identical observations at the start of episodes.

## Performance
On my MacBook Pro (Retina, 15-inch, Mid 2015), running
```
python train.py --num-workers 6 --env-id Pong-v0 --log-dir /tmp/pong
```
I get about 80 fps:
![image](https://cloud.githubusercontent.com/assets/219508/21966992/eb529982-db32-11e6-802c-bd1e6727e939.png)

This is with Anaconda python 3.5, tensorflow 0.11.0 and OpenCV 3.1.0. I get similar results with tf 0.12.1. OpenCV uses a lot of the CPU time, and different versions use the hardware in different ways. If you're using install from the starter-agent README, it should be fine. If you're using another install, see if there are installation options around how it parallelizes. M

**Comment by tr1pzz** [comment]
Whuuuuut, I just found my error loool!

I trained the Agent on "PongDeterministic-v3" but in my visualization script I used "Pong-v0"...
So apparently some settings are slightly different, completely wrecking the Agent's behavior.

Well, anyway thx a lot for your fast feedback, it is much appreciated.
Now I can finally start trying some actual changes to the model :)
Are there any good fora where people are discussing approaches they are trying or should I stick to reading papers etc.. ?

best regards,
I guess this topic can be closed :+1:
